File: xkcd/src/main/java/xyz/jienan/xkcd/model/QuoteModel.java
package xyz.jienan.xkcd.model;

import java.util.Random;

import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;
import xyz.jienan.xkcd.BuildConfig;
import xyz.jienan.xkcd.base.network.NetworkService;
import xyz.jienan.xkcd.base.network.QuoteAPI;

public class QuoteModel {

    private static final long INTERVAL = BuildConfig.DEBUG ? 10000 : 1000 * 60 * 60 * 24;

    private static QuoteModel quoteModel;

    private final QuoteAPI quoteAPI = NetworkService.getQuoteAPI();

    private QuoteModel() {
        // no public constructor
    }

    public static QuoteModel getInstance() {
        if (quoteModel == null) {
            quoteModel = new QuoteModel();
        }
        return quoteModel;
    }

    public Observable<Quote> getQuoteOfTheDay(final Quote previousQuote) {
        return Observable.just(1)
                .flatMap(ignored -> {
                    if (System.currentTimeMillis() - previousQuote.getTimestamp() < INTERVAL) {
                        return Observable.just(previousQuote);
                    } else {
                        return queryNewQuote(previousQuote);
                    }
                });
    }

    private Observable<Quote> queryNewQuote(final Quote quote) {
        return quoteAPI.getQuotes()
                .subscribeOn(Schedulers.io())
                .doOnNext(quotes -> quotes.remove(quote))
                .map(quotes -> quotes.get(new Random().nextInt(quotes.size())))
                .doOnNext(result -> result.setTimestamp(System.currentTimeMillis()));
    }
}

File: xkcd/src/main/java/xyz/jienan/xkcd/model/QuoteModel.kt
package xyz.jienan.xkcd.model

import android.content.res.Resources
import androidx.annotation.RawRes
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import io.reactivex.Observable
import io.reactivex.schedulers.Schedulers
import xyz.jienan.xkcd.BuildConfig
import xyz.jienan.xkcd.R
import xyz.jienan.xkcd.base.network.NetworkService
import java.util.*

object QuoteModel {

    private val INTERVAL = (if (BuildConfig.DEBUG) 10000 else 1000 * 60 * 60 * 24).toLong()

    fun getQuoteOfTheDay(previousQuote: Quote, resources: Resources): Observable<Quote> =
            Observable.just(previousQuote)
                    .flatMap {
                        if (System.currentTimeMillis() - it.timestamp < INTERVAL) {
                            Observable.just(it)
                        } else {
                            queryNewQuote(it, resources)
                        }
                    }

    private fun queryNewQuote(quote: Quote, resources: Resources) =
            NetworkService.quoteAPI.quotes
                    .subscribeOn(Schedulers.io())
                    .onErrorReturnItem(resources.readRawJson(R.raw.quotes))
                    .map { quotes ->
                        quotes.remove(quote)
                        quotes
                    }
                    .map { quotes -> quotes[Random().nextInt(quotes.size)] }
                    .doOnNext { result -> result.timestamp = System.currentTimeMillis() }
}

private inline fun <reified T> Resources.readRawJson(@RawRes rawResId: Int): T {
    openRawResource(rawResId).bufferedReader().use {
        return Gson().fromJson<T>(it, object: TypeToken<T>() {}.type)
    }
}